File: runtime/src/aco_runtime_lib/prompts/renderer.py
"""Strict, deterministic prompt renderer.

Same semantics as `packages/prompts/src/renderer.ts`. Variables
must be present in the context; missing variables raise.
"""

from __future__ import annotations

import re
from typing import Any

_VAR_PATTERN = re.compile(r"\{\{\s*([a-zA-Z_][\w.]*)\s*\}\}")


class PromptRenderError(Exception):
    """Raised when a variable is missing or of an unsupported type."""

    def __init__(self, message: str, path: str) -> None:
        super().__init__(f"{path}: {message}")
        self.path = path


def _lookup(ctx: dict[str, Any], path: str) -> str:
    cur: Any = ctx
    for part in path.split("."):
        if not isinstance(cur, dict):
            raise PromptRenderError(f"cannot read {part!r} of non-object", path)
        if part not in cur:
            raise PromptRenderError("undefined variable", path)
        cur = cur[part]
    if cur is None:
        return "null"
    if isinstance(cur, (str, int, float, bool)):
        return str(cur)
    raise PromptRenderError(f"unsupported type: {type(cur).__name__}", path)


def render(template: str, ctx: dict[str, Any]) -> str:
    """Render `template` against `ctx`."""
    return _VAR_PATTERN.sub(lambda m: _lookup(ctx, m.group(1)), template)


def estimate_tokens(text: str) -> int:
    """Rough token estimate (4 chars ≈ 1 token)."""
    return (len(text) + 3) // 4

File: packages/prompts/src/renderer.ts
/**
 * Minimal prompt template renderer.
 *
 * Supports `{{ var.path }}` substitution with strict, deterministic
 * semantics. **No** `eval`, **no** arbitrary expressions. Variables
 * must be present in the context; missing variables throw.
 *
 * Mirrors the Python renderer in `runtime/prompts/renderer.py`. The
 * two implementations are kept in sync manually; CI runs a fixture
 * diff to catch drift.
 */

export class PromptRenderError extends Error {
  constructor(
    message: string,
    public readonly path: string,
  ) {
    super(`${path}: ${message}`);
    this.name = 'PromptRenderError';
  }
}

const VAR_PATTERN = /\{\{\s*([a-zA-Z_][\w.]*)\s*\}\}/g;

export type Context = Readonly<Record<string, unknown>>;

function lookup(ctx: Context, path: string): string {
  const parts = path.split('.');
  let cur: unknown = ctx;
  for (const p of parts) {
    if (cur === null || cur === undefined || typeof cur !== 'object') {
      throw new PromptRenderError(`cannot read "${p}" of non-object`, path);
    }
    cur = (cur as Record<string, unknown>)[p];
    if (cur === undefined) {
      throw new PromptRenderError(`undefined variable`, path);
    }
  }
  if (cur === null) return 'null';
  if (typeof cur === 'string') return cur;
  if (typeof cur === 'number' || typeof cur === 'boolean') return String(cur);
  throw new PromptRenderError(`unsupported type: ${typeof cur}`, path);
}

export function render(template: string, ctx: Context): string {
  return template.replace(VAR_PATTERN, (_match, path: string) => lookup(ctx, path));
}

/** Estimate token count (very rough: 4 chars ≈ 1 token). */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}
